[WIP] Реалізація міграцій БД з використанням doctrine migrations

yukoff opened the pull request:
Щоб в майбутньому бути сумісним з laravel (upstream TP перейшов на нього в experimental
гілці) почнімо з міграцій doctrine.

Через використання  ENUM довелось задіяти ORM та згенерувати відповідні entities, плюс необхідно проаналізувати, чи можливий т.зв. data-seeding у випадку doctrine (якщо замість альтерів таблиць з даними переливати дані зі старої структури в нову, пусту базу)

- [x] ~~cli-додаток для роботи з міграціями~~ - використовується `artisan` з `laravel` (upstream переходять на нього)
- [x] міграція з TP v2.2.1 на v2.2.2, якщо контейнер вже оновлювався - міграція не повинна "падати"
- [x] конвертація бази в `utf8mb4`/`utf8mb4_unicode_ci`
- [x] кастомні типи даних doctrine для ENUM типу даних
- [x] використання `utf8mb4_bin` для зберігання паролів, ip-адрес, ідентифікаторів сесій тощо - частково це було реалізовано до пропонованих змін
- [x] ~~проаналізувати можливість data-seeding у випадку doctrine~~ - `laravel` може сідування даних
- [x] спробувати замість класів з анотаціями метадані зберігати в `.yml`
- [x] змінити простір імен для класів ORM, налаштувати їх автозавантаження в `composer.json`

Closes #7, closes #46, closes #76 

## TODO

- [x] Синхронізація бази з моделями doctrine
- [ ] додати в `docker-compose.yml`, щоб міграція автоматично запускалася при `docker-compose up -d` (документація?)
  - [ ] Wrapper для `php artisan migrate`
- [x] Змержити з `experimental` гілки [upstream](/torrentpier/torrentpier/tree/experimental)’у правки щодо зміни конфліктуючих з laravel імен функцій (`redirect()` & `auth()`)
- [x] Злити PR міграції зі старої толоки, closes #76 (in progress)
  - [x] `bb_attachments` (Прикріплення файлів до приватних повідомлень) #27 
  - [x] `bb_attachments_rating` (кнопка "Дякую") #23/#53
  - [x] `bb_bt_config`, closes #95
  - [x] `bb_bt_search_results`, closes #94
  - [x] `bb_bt_torrents`.`topic_check*`, closes #43
  - [x] `bb_bt_torrents_del`, closes #96 
  - [x] `bb_bt_users`, closes #47
  - [x] `bb_bt_users_dl_status`, closes #97 
  - [x] `bb_confirm`, closes #98 
  - [x] `bb_dl_st_prevs_list`, closes #99 
  - [x] `bb_easymod` та `bb_easymod_processed_files`, closes #39 
  - [x] `bb_flags`, closes #100 
  - [x] `bb_forbidden_extensions`, closes #101 
  - [x] `bb_forum_prune`, closes #102 
  - [x] `bb_posts_edit`, closes #40 
  - [x] `bb_search_wordlist`, closes #103 
  - [x] `bb_search_wordmatch`, closes #90
  - [x] `bb_sessions_keys`, closes #37 
  - [x] `bb_themes` та `bb_themes_name`, closes #104 
  - [x] `bb_topics_move`, closes #105 
  - [x] `bb_user_group.group_moderator`, closes #38 
  - [x] `bb_vote_(desc|results|voters)`, closes #91 
  - [x] `bb_warnings` та `bb_warnings_text` (так звані "жовті картки") #41 
- [x] Прибрати тригер для анонсера XBT, closes #108
- [x] Оновити паролі користувачів на тестових інсталяціях відповідно до #50, closes #54 
- [ ] Мінімізувати розбіжності у полях типу `TEXT`, які при `CONVERT TO CHARACTER SET` стають `MEDIUMTEXT` типу
- [ ] Імпорт даних для нових таблиць при міграції зі старої версії, closes #106
- [ ] Типові значення стовпців для `bb_users`, closes #56 
- [ ] PRIMARY KEY на таблиці bb_bt_dlstatus #107

yukoff committed:
Implement database migrations (doctrine)

To be compatible with laravel (upstream TP switch to it in experimental
branch) we'd start with doctrine migrations.

* Add simple db console with migrations and orm commands exposed.
* Adapt for upstream laravel structure

konfuciusu commented:
> використання utf8mb4_bin для зберігання паролів, ip-адрес, ідентифікаторів сесій тощо - частково це було реалізовано до пропонованих змін

Добре, що не набули про це при переході на unicode_ci :+1:

yukoff committed:
[db] Remove odd default values for PK columns

Also make sure user_id is unsigned across all tables

yukoff committed:
Add .htaccess config for new files

konfuciusu commented:
Наївна реалізація завдання про `docker-compose`:

```
toloka-migration:
    image: hurtom/php:nginx
    command: bash -c "sleep 30 && doctrine migration command"
    volumes:
      - ./:/var/www/html
    env_file:
      - dev.env
```

На жаль, автори docker відкинули docker/compose#374 і docker/compose#1809, тому напевне єдина альтернатива - це додати `entrypoint`-скрипт, але тоді варто звернути увагу на docker/compose#3140

konfuciusu commented:
Можна ще за допомогою [`supervisord`](https://github.com/hurtom/php/blob/master/7.1/nginx/supervisord.conf):

```
[program:migration]
command = doctrine migration command
autorestart = false
stdout_logfile = /dev/stdout
stdout_logfile_maxbytes = 0
stderr_logfile = /dev/stderr
stderr_logfile_maxbytes = 0
```

yukoff commented:
Про `supervisord` я й не подумав. В принципі, мені головне включити їх до процесу тестування, та й +1 рядок в документації теж не проблема :) Можна мабуть через окремий issue/PR. Зараз аналізую/адаптую код з #76, перебазований над цим - все ніяк до docker-compose не доберусь.

konfuciusu commented:
Зараз ще згадав: `doctrine migration command` бажано загорнути в невеликий скрипт (php/bash), який запускає комадну аж тоді, коли mariadb стане доступною.

yukoff committed:
Add new entities from old schema, review existing

yukoff committed:
Add new mappings from migrated from old schema

yukoff committed:
Update entities acording to updated mappings

yukoff committed:
Add fulltext flags for FULLTEXT indexes etc

yukoff committed:
Delay index creation after changing charset for tables

yukoff committed:
Add BBCodeHelper etc

yukoff committed:
Resolve conflicting function names

This is partial sync with torrentpier/experimental, required to resolve
conflicts with laravel helper functions

yukoff committed:
[db] Migration of bb_posts*

konfuciusu commented:
Таблиці, які не позначані `***`, можна пропустити:

```
bb_bt_config
bb_bt_search_results
bb_bt_torrents_del
bb_confirm
bb_dl_st_prevs_list
bb_easymod
bb_easymod_processed_files
bb_flags
bb_forbidden_extensions
bb_forum_prune
bb_search_wordlist
bb_search_wordmatch
bb_sessions_keys
bb_themes
bb_themes_name
bb_topics_move
```

Вони не мають жодної цінності.

yukoff commented:
@konfuciusu Трохи не вловив - а позначені де? Тобто цей перелік
```
bb_bt_config
bb_bt_search_results
bb_bt_torrents_del
bb_confirm
bb_dl_st_prevs_list
bb_easymod
bb_easymod_processed_files
bb_flags
bb_forbidden_extensions
bb_forum_prune
bb_search_wordlist
bb_search_wordmatch
bb_sessions_keys
bb_themes
bb_themes_name
bb_topics_move
```
можна викидати?

konfuciusu commented:
Тут: https://github.com/hurtom/toloka/blob/master/install/sql/toloka_old.sql

Так, можна викидати.

yukoff committed:
[db] Table bb_bt_users should not use AI for user_id

yukoff commented:
@konfuciusu Аааа... Вибачаюсь, я не звернув увагу на ті "зірочки" в старій схемі 🤦‍♂️ 😄

yukoff committed:
[db] SQL formatting to simplify code review

yukoff committed:
[db] Get rid of unused bb_bt_config table (#95)

yukoff committed:
[db] Get rid of unused bb_bt_search_results table (#94)

yukoff committed:
[db] Get rid of unused bb_bt_torrents_del table (#96)

yukoff committed:
[db] Get rid of unused bb_confirm table (#98)

yukoff committed:
[db] Get rid of unused bb_dl_st_prevs_list table (#99)

yukoff committed:
[db] Get rid of unused bb_easymod(_processed_files)? table (#39)

yukoff committed:
[db] Get rid of unused bb_flags table (#100)

yukoff committed:
[db] Get rid of unused bb_forbidden_extensions table (#101)

yukoff committed:
[db] Get rid of unused bb_forum_prune table (#102)

yukoff committed:
[db] Get rid of unused bb_search_wordlist table (#103)

yukoff committed:
[db] Get rid of unused bb_search_wordmatch table (#90)

yukoff committed:
[db] Get rid of unused bb_sessions_keys table (#37)

yukoff committed:
[db] Get rid of unused bb_themes(_name)? tables (#104)

yukoff committed:
[db] Get rid of unused bb_topics_move table (#105)

yukoff committed:
[db] Add migration of bb_bt_users_dl_status (#97)

yukoff committed:
Phinx is not required for migrations

yukoff committed:
[db] Get rid of unused trigger bb_topics_au (#108)

yukoff commented:
@konfuciusu Я [тут](https://github.com/hurtom/toloka/issues/91#issuecomment-319231534) питав, чи можливо отримати вибірку/дамп для трьох таблиць для якогось опитування (стосовно #91):
* `bb_vote_desc`
* `bb_vote_results`
* `bb_vote_voters`

Можна `mysqldump`-ом зробити, приміром
```
mysqldump ... --where="vote_id=X"
```

Заодне це дозволить зробити тести

yukoff committed:
[db] Migrate bb_vote_(desc|results|voters) data (#91)

yukoff committed:
[db] Adjust PK and indexes on bb_posts_edit (#40)

yukoff committed:
[db] Add bb_attachments_rating to migration (#23)

This one is to reflect changes from PR #53 and to be applied during
database migration

yukoff committed:
[db] Add topic_check* migrations and optimizations

* Add migration for bb_bt_torrents.topic_check* (#43)
* Aggregate ALTERs to reduce table modifications

yukoff committed:
[db] Attachments to private messages (#27)

yukoff committed:
[db] Format source and adjust code

* Source formatting, reordered steps alphabetically
* Fetch via iterator instead of fetchAll()
* Convert rest of tables to InnoDB

yukoff committed:
[db] Adjust to review comments in #53

yukoff committed:
[db] Fix varbinary length, improve bb_bt_tor* migration

yukoff committed:
[db] Fix incomplete (unportable) queries etc

TIMESTAMP fields should have set {version: true} in field description.
Also make corresponding fields timestampable via gedmo/doctrine-extensions.
Make quoting ENUM values the right way

yukoff committed:
[db] Drop missed torrent_id after migration (#43)

yukoff committed:
[db] Adjust PK for bb_attachments (#27)

yukoff committed:
[db] Minimize diff, comply with doctrine types

* Reorder clauses for readability
* Adjust cat_id datatype, bb_categories and related tables
* Adjust forum_id datatype
* Adjust topic_id and related
* Adjust post_id and related
* Adjust group_id and related
* Adjust warning_id and related

yukoff committed:
[db] Add column for thanks button during migration (#23)

Follow up review comments in PR #53, add `thanks_count` column

yukoff committed:
[db] Support either way of update

Migration should succeed with either way of migration - from the old 0.x
engine to current or from the stock 2.x schema.

Reset some table options, eg. MAX_ROWS and ROW_FORMAT.

Reflect changes for TP v2.2.0 (at this point all `*_ip` columns were
reset to '0', drop bb_ads table etc) and v2.2.2

Adjust datatypes to comply with doctrine etc

yukoff committed:
Follow-up #111 - update requirements etc

* Update `codeception/c3`
* Update `laravel-doctrine/extensions` (to min. version with support for
  laravel 5.4)
* Update `laravel-doctrine/migrations` (to min. version supporting
  laravel 5.4)
* Update `laravel-doctrine/orm`
* Add redis sample config (.env)
* Use `file` as a default cache driver

yukoff committed:
[db] Update stock user passwords for #50

yukoff commented:
Я тут думаю - може варто замість автоматичного запуску через `docker-compose up -d` все ж робити це за допомогою composer
```diff
diff --git a/composer.json b/composer.json
index ff0de5d2..ecdfdcfa 100644
--- a/composer.json
+++ b/composer.json
@@ -88,6 +88,9 @@
       "Illuminate\\Foundation\\ComposerScripts::postUpdate",
       "php artisan optimize"
     ],
+    "database:migrate": [
+      "php artisan doctrine:migrations:migrate --force"
+    ],
     "test": [
       "echo 'NOTE: Instead of running <composer test:lint> linting should be done with StyleCI'",
       "composer test:unit",
```

та відповідно
```shell
docker-compose up -d
docker-compose exec -u www-data toloka composer update -o
docker-compose exec -u www-data toloka composer database:migrate
```
Або можна навіть `database:migrate` додати у `post-update-cmd`, тоді остання команда не потрібна

yukoff committed:
[db] Add database:migrate composer command

Add `database:migrate` composer command and run it as part of travis `before_script`
Also use username instead of user id ("magic number")

yukoff committed:
[db] Merge remaining queries from #76 etc

* Merge remaining queries, reorder and add references to issues/PRs
* Adjust datatype for topic_type
* Add validation for torrent check status
* Some source formatting etc

konfuciusu commented:
Через `composer database:migrate` теж нормально, треба тільки це тоді додати в readme.

Можна вже пробувати тестувати на старій базі?

yukoff commented:
@konfuciusu Майже, є кілька моментів які зловив при перебазуванні #55 - днями з ними розберусь, думаю десь на наступному тижні вже можна буде тестувати.

В принципі міграція працює вже зараз, але після міграції не все працює з функціоналу через нові поля і відсутність типових значень на них. Зараз можна хіба що прогнати тестову міграцію на знятку (снепшоті), аби прикинути необхідний час і можливо щось оптимізувати.
